design-documents/testing-strategy.md

```diff
@@ -0,0 +1,39 @@
+
+Despite running 20,000 - 30,000 tests per pull request, we have little to no confidence in the stability of the product. This means that we still need a manual regression process that lasts weeks in the best of cases. 
+
+A sizable portion of this can be attributed to our current high-level testing mentality that prioritizes system-level black box testing over other types of tests. While these types of tests are necessary and good to have, they only provide implicit coverage to the code paths of the concrete implementations that happen to be loaded in the high-level black box tests (they are also expensive/difficult to maintain). We still need explicit coverage for the concrete implementations that extension developers have hard dependencies on. This is true for the simple code paths as well as variations both positive and negative flows.
+
+### Design
+
+#### New Testing Strategy:
+
+The new strategy for testing relies on having test coverage at each layer. The idea is to have many small and fast tests. Therefore, tests MUST be written in the following order of granularity and the higher-level the test, the fewer tests there will be:
+
+1. Code MUST be tested on its own. This MAY be done through unit or integration testing as appropriate. This level of test coverage ensures that the code can properly function on its own.  
+1. Code MUST be shown to work with known collaborators. This MUST be done through integration tests and ideally SHOULD involve as few components as possible per test case. This level of tests ensures that the code will work with its known associates. Examples of such associates would include aggregators, composites, etc.
+1. webapi behavior MUST be covered by blackbox tests to assert that the service contracts are properly satisfied. This ensures that the software package as a whole functions properly when everything is compiled together. This includes REST/SOAP/GraphQl etc. To be clear, these tests do not exclude the underlying concrete implementations from being tested. 
+1. MFTF tests SHOULD only be used to cover P0/P1 testing scenarios. Until these scenarios are all completed, a product owner MUST be consulted for approval of new tests. Additionally, MFTF MUST NOT be used for testing scenarios that are not P0/P1
+
+In addition, regardless of the interpretation of the strategy above: Any class marked with an `@api` annotation MUST be covered with explicit test coverage via integration or unit tests. These tests MUST test the concrete implementation's behavior in a way that can not be inadvertently changed outside of the test itself. For example: testing a concrete class through DI preference, is not safe because it could be overridden via configuration of another loaded module.
+
+#### What to test when changing code
+1. Before changing code or writing new code, write passing tests (either blackbox or whitebox as appropriate) for existing behavior that doesn't already have test coverage.
```

Comment:
What would be an example of white box test? Integration and functional tests generally considered blackbox tests as far as I know. A unit test? Writing white box unit test will not help much as it will break when you start making changes (including those that don't affect functionality). I suggest to clarify that we want to write black box tests in general and white box tests only to test SPI. And instead of black and white box tests here I would just name different types of tests unit, integration, API and functional.

Reply:
Again, referring to my previous comments above, there are times when black box testing doesn't make any sense. For example, if you have a method signature like `filter($value):string` the contract is pointless to test totally blackbox in the sense that you can't simply verify basic input=>output. You should definitely do that but it isn't enough in most cases. The implementation will dictate how complicated the testing scenarios will be. 

What seems to be the root issue here is there is a Magento-specific terminology for what "whitebox test" means. For Magento this seems to be defined as test that is a direct mirror of the code. But that isn't true, a whitebox test would be one where it has some level of knowledge about the implementation. Such cases are good for making sure the implementation handles corner cases (unique to the implementation) correctly.

Again, the idea is that contracts are great but there are still concrete dependencies on implementations and implementation can have individualized flaws/weaknesses specific to the details of that implementation.